### DISCUSSION TITLE
Any ideas for 3D text generation and rendering

### DISCUSSION THREAD

**Discussion by cocor-au-lait** [needle-tools/needle-engine-support#79] [Q&A] [answered]
Hello, I'm interested in this project.

I'm trying to generate 3d text mesh in AR space dynamically.

Needle Engine supports UI Text, but I'd like to express stereoscopic rich text.

I tried to use TextMesh Pro on Unity but it didn't seem to be supported in this project.

https://github.com/needle-tools/needle-engine-support/issues/70

From the issues mentioned it doesn't seem possible to interactively change the text component.

So is there a way to create a static text mesh from any strings?

**Comment by marwie** [comment, marked as answer]
Hey, 

you can use a worldspace canvas with the "old" Text component. Altough it does not support RichText in the current version right now (it did in the past and the code for it is still there but was temporarely disabled for automatically generating font atlases from Unity).

With the classic Text component text [can be changed dynamically](https://twitter.com/marcel_wiessler/status/1574704684354490369)

You could also write a custom component and using some of the [core threejs methods for generating text](https://threejs.org/docs/#manual/en/introduction/Creating-text).  

Also have a look at the [samples repository](https://github.com/needle-tools/needle-engine-samples) which has a [basic 3d mesh component](https://github.com/needle-tools/needle-engine-samples/blob/main/package/Runtime/Needle%20Engine%20Samples~/TextGeometry.ts) too that might help in getting started

**Comment by cocor-au-lait** [comment]
@marwie Thanks for the quick reply 🚀

I tried importing code from [TextGeometry.ts](https://github.com/needle-tools/needle-engine-samples/blob/main/package/Runtime/Needle%20Engine%20Samples~/TextGeometry.ts) into my project and succeeded in rendering 3D text that could be changed at runtime.

<img width="309" alt="image" src="https://user-images.githubusercontent.com/30252919/193425380-d2d1cff9-1bc1-49f1-8dd7-7418a053dc77.png">

Apparently the font path seems to be hard-coded, but it seems possible for using any font to display specific character by defining serializeable field.

https://github.com/needle-tools/needle-engine-samples/blob/36e7322f55b1251b2ca1631466c32d64e54cd100/package/Runtime/Needle%20Engine%20Samples~/TextGeometry.ts#L42

Any plans to include this component in future releases?

**Reply by marwie** [reply]
Hey yes we will have support for that / TextMeshPro support in a future release (just no ETA at the moment)

With the newer version you can change your C# component stub to use a font asset instead too - this will in turn produce a font atlas which you should be able to use in your typescript component. 

You can have a look at how the ``Text`` component does it in Typescript (by clicking the blue link)

![image](https://user-images.githubusercontent.com/5083203/193426153-3fccd7a7-079c-4885-9f33-890c8070c612.png)

**Reply by cocor-au-lait** [reply]
Thanks, I'm looking forward to future releases.

**Reply by cocor-au-lait** [reply]
@marwie Hi, it's been a while.
Is it still undecided when `TextMeshPro` will be supported?

**Comment by cocor-au-lait** [comment]
[Bruno Simon 20k](https://bruno-simon-20k-needle.glitch.me/) included in [Needle Engine Samples](https://github.com/needle-tools/needle-engine-samples) appeared to rendering TextGeometry but was actually just static mesh (glb) :eyes:

I think that the introduced method can be replaced with TextGeometry, so I will try it.

**Reply by marwie** [reply]
Yes that's a static mesh since it just uses the assets provided by Bruno
